### DISCUSSION TITLE
What's the right way to call an Endpoint method from a WidgetRoute implementation hosted in the web server of the same app ?

### DISCUSSION THREAD

**Discussion by rotoclabs** [serverpod/serverpod#2519] [Q&A] [answered]
I get the idea that we use the generated client library  in the actual client application, like the flutter app. My situation needs a classic Rest API response delivered via something like `WidgetJson` due to a need of supporting 3rd party legacy client software.

I do understand that I can access DB through the passed session object, in several of the web server handlers. I could obviously do what I need through that. But a lot of my CRUD logic is pretty well implemented in the Endpoint. Now how do I benefit from that ? Is it that you do the same `await client.notes.createNote(noteObjFromRestBody)` from the `build()` of the `WidgetRoute` implementation, and if so how and where do I get the client object ? I would like to know if there is more direct way to do this, since this not an external client. Is there some way to get access to the instance of the Endpoint directly to make this call ?

I think this question might be related to another one which I found in the discussion, asking if one could call an endpoint from another endpoint.

*** Update ***

I've done this in a bit of a roundabout way.

In server.dart
```
 pod.webServer.addRoute(
      NotesApiRoute(
        endpoints: pod.endpoints,
      ),
      '/api/notes');
```

and in the notes_api_route.dart

```
class NotesApiRoute extends WidgetRoute {
  final EndpointDispatch endpoints;

  NotesApiRoute({required this.endpoints});

  @override
  Future<bool> handleCall(Session session, HttpRequest request) async {
      // stuff, more stuff

      await (endpoints.getConnectorByName('note')!.endpoint as NoteEndpoint)
          .createNote(session, theNote);

      // resolve widget
      // render etc.
  }
}
```

It worked for me. But is this a good approach  ?

**Comment by SandPod** [comment, marked as answer]
Hey @robythomas 👋 

I would recommend moving the logic that you would like to share out from the endpoint and then make both the endpoint and the relic route call that same place.

There is no need for this logic to be defined solely in an endpoint, and if it is shared between multiple places it your code it also makes it clear that the logic is not solely connected to an RPC call from the client.
